### PR TITLE
core(config): special case full-page-screenshot audit in filtering

### PR DESCRIPTION
**Pull request by connorjclark**
not pretty. potentially better than alternative (?), which is to ditch the FPS audit and make it a property of the LHR... (`lhr.fullPageScreenshot`).

**why**: CDT uses `onlyCategories`, which means this audit is always skipped because it belongs to no category :(

### CODE REVIEW

**Comment by patrickhulce** [comment]
Just throwing this alternative out there, put it in all of them? :)

**Comment by brendankenny** [comment]
> potentially better than alternative (?), which is to ditch the FPS audit and make it a property of the LHR... (`lhr.fullPageScreenshot`).

The property could go wherever feels good. The reasoning is that the desired behavior is basically exactly the same as stackPacks. Gathered and audited regardless of which audits are run or filtered out, and the results aren't used as an audit in the report, they're used to provide extra information in a bunch of other audits.

> Just throwing this alternative out there, put it in all of them just hidden? :)

We could revive @adrianaixba's hidden audit PR :)

File: lighthouse-core/config/config.js [hunk 1]
```diff
@@ -656,6 +656,19 @@ class Config {
       }
     });
 
+    // The `full-page-screenshot` audit belongs to no category, but we still want to include
+    // it for the categories that utilize it.
+    if (settings.onlyCategories) {
```
Comment (adamraine) [inline lighthouse-core/config/config.js:661]: What if `onlyAudits` is set? In that situation, `full-page-screenshot` should be explicitly listed in `onlyAudits` right?

**Comment by connorjclark** [comment]
> We could revive @adrianaixba's hidden audit PR :)

I'd prefer this, but how much work does that entail? If we are pressed for time this PR seems fine to me, for now.

File: lighthouse-core/config/config.js (hunk 1 above)
Comment (connorjclark) [reply lighthouse-core/config/config.js:661]: yeah that's my understanding

**Comment by patrickhulce** [comment]
> I'd prefer this, but how much work does that entail? If we are pressed for time this PR seems fine to me, for now.

Probably not worth it, IMO, but seemed like you were searching for alternatives :) 

I'd say we can just explore the `lhr.fullPageScreenshot` style for v8 and this works until then.

**Comment by benschwarz** [comment]
We use `onlyAudits` and `onlyCategories` quite a bit. FPS is one of the audits that's regularly skipped by our test verification/scheduler as it caused `PROTOCOL_TIMEOUT` or crashed Chrome for certain sites. 

Are you saying that it'd not be configurable to opt out of it running?

**Comment by patrickhulce** [comment]
@benschwarz you can still use `skipAudits` to opt-out of it

**Comment by paulirish** [comment]
here i wrote you a test :)  


```js


    it('should keep uncategorized audits even if onlyCategories is set', () => {
      assert.ok(origConfig.audits.includes('full-page-screenshot'));
      // f-p-s is non-categorized
      const matchCategories = Object.values(origConfig.categories).filter(cat =>
          cat.auditRefs.find(ref => ref.id === 'full-page-screenshot'));
      assert.equal(matchCategories.length, 0);


      const extended = {
        extends: 'lighthouse:default',
        settings: {
          onlyCategories: ['accessibility'],
        },
      };
      const config = new Config(extended);

      assert.ok(config.audits.find(a => a.path.includes('full-page-screenshot')));
    });
```

File: lighthouse-core/test/config/config-test.js [hunk 2]
```diff
@@ -663,8 +663,11 @@ describe('Config', () => {
     });
 
     assert.ok(config.audits.length, 'inherited audits by extension');
-    assert.equal(config.audits.length, origConfig.categories.performance.auditRefs.length + 1);
+    // +1 for `is-on-https`, +1 for `full-page-screenshot`.
+    assert.equal(config.audits.length, origConfig.categories.performance.auditRefs.length + 2);
     assert.equal(config.passes.length, 1, 'filtered out passes');
+    assert.ok(config.audits.filter(a => a.implementation.meta.id === 'is-on-https'));
```
Comment (adamraine) [inline lighthouse-core/test/config/config-test.js:669]: `assert.ok([])` does not fail but this is the intent correct?

Comment (connorjclark) [reply lighthouse-core/test/config/config-test.js:669]: ooooooooooooooooooo shiet. good catch!

File: lighthouse-core/test/config/config-test.js [hunk 3]
```diff
@@ -1165,9 +1170,12 @@ describe('Config', () => {
       };
       const config = new Config(extended);
       const selectedCategory = origConfig.categories.performance;
-      const auditCount = Object.keys(selectedCategory.auditRefs).length + 1;
+      // +1 for `works-offline`, +1 for `full-page-screenshot`.
```
Comment (adamraine) [inline lighthouse-core/test/config/config-test.js:1173]: nit: Should `works-offline` be `service-worker` instead?

**Review by adamraine — APPROVED** [review]
Overall LGTM

File: lighthouse-core/test/config/config-test.js (hunk 3 above)
Comment (connorjclark) [inline lighthouse-core/test/config/config-test.js:1173]:
```suggestion
      // +1 for `service-worker`, +1 for `full-page-screenshot`.
```

**Comment by paulirish** [comment]
coolio. @brendankenny is gonna write up an issue exploring our options for where we can take this.

**Pull request merged**